include/material/norsand.tcc

```diff
@@ -0,0 +1,693 @@
+                2 * std::pow(dF_dsigma(5), 2));
+
+  // Compute hardering term
+  double hardening_term;
+  if (bond_model_) {
+    // Derivatives in respect to p_cohesion
+    const double dF_dpcohesion =
+        M_theta / N_ *
+        (1 +
+         (N_ - 1) * std::pow(((mean_p + p_cohesion) /
+                              (p_image + p_cohesion + p_dilation)),
+                             (N_ / (1 - N_))) -
+         N_ * (p_image + p_dilation - mean_p) /
+             (p_image + p_cohesion + p_dilation) *
+             std::pow(
+                 ((mean_p + p_cohesion) / (p_image + p_cohesion + p_dilation)),
+                 (N_ / (1 - N_))));
+
+    const double dpcohesion_depsd =
+        -p_cohesion_initial_ * m_cohesion_ *
+        exp(-m_cohesion_ * (*state_vars).at("epds"));
+
+    // Derivatives in respect to p_dilation
+    const double dF_dpdilation =
+        -1. * M_theta *
+        std::pow(((mean_p + p_cohesion) / (p_image + p_cohesion + p_dilation)),
+                 (1 / (1 - N_)));
+
+    const double dpdilation_depsd =
+        -p_dilation_initial_ * m_dilation_ *
+        exp(-m_dilation_ * (*state_vars).at("epds"));
+
+    hardening_term = dF_dpi * dpi_depsd * dF_dsigma_deviatoric +
+                     dF_dpcohesion * dpcohesion_depsd * dF_dsigma_deviatoric +
+                     dF_dpdilation * dpdilation_depsd * dF_dsigma_deviatoric;
+  } else {
+    hardening_term = dF_dpi * dpi_depsd * dF_dsigma_deviatoric;
+  }
+
+  // Construct Dp matrix
+  this->dp_ = (de_ * dF_dsigma * dF_dsigma.transpose() * de_) /
+              (dF_dsigma.transpose() * de_ * dF_dsigma - hardening_term);
+
+  return true;
+}
+
+//! Compute stress
+template <unsigned Tdim>
+Eigen::Matrix<double, 6, 1> mpm::NorSand<Tdim>::compute_stress(
+    const Vector6d& stress, const Vector6d& dstrain,
+    const ParticleBase<Tdim>* ptr, mpm::dense_map* state_vars) {
+
+  // Note: compression positive in all derivations
+  Vector6d stress_neg = -1 * stress;
+  Vector6d dstrain_neg = -1 * dstrain;
+
+  // Compute current stress invariants (need to remove this, only needed in the
+  // first step)
+  this->compute_stress_invariants(stress_neg, state_vars);
+
+  // Elastic step
+  // --------------------------------------------------------------------------------------
+  // Using critical state volumetric
+  // Bulk modulus
+  bulk_modulus_ =
+      (1. + (*state_vars).at("void_ratio")) / kappa_ * (*state_vars).at("p");
+  // Shear modulus
+  if (bond_model_) {
+    shear_modulus_ = 3. * bulk_modulus_ * (1. - 2. * poisson_ratio_) /
+                         (2.0 * (1. + poisson_ratio_)) +
+                     m_shear_ * ((*state_vars).at("p_cohesion") +
+                                 (*state_vars).at("p_dilation"));
+  } else {
+    shear_modulus_ = 3. * bulk_modulus_ * (1. - 2. * poisson_ratio_) /
+                     (2.0 * (1. + poisson_ratio_));
+  }
+
+  // Using Young's Modulus
```

Comment:
Could you please remove these comments, if you are not using those?

Reply:
OK. I initially wanted different definition of those. But it's ok, let's stick with one.